### PR TITLE
feat: add GOOGLE_WORKSPACE_CLI_API_BASE_URL for custom/mock endpoint support

### PR DESCRIPTION
**Pull request by xdotli**
## Reproduce steps:

```bash
# Install from this branch
cargo install --git https://github.com/benchflow-ai/cli --branch feat/custom-endpoint-override
```

**Without env var** — request goes to Google APIs:

```bash
$ gws gmail users getProfile --params '{"userId":"me"}' --dry-run
{
  "method": "GET",
  "url": "https://gmail.googleapis.com/gmail/v1/users/me/profile"
}
```

**With `GWS_API_BASE_URL`** — request redirected to custom endpoint:

```bash
$ GWS_API_BASE_URL=http://localhost:8001 gws gmail users getProfile --params '{"userId":"me"}' --dry-run
[gws] Custom API endpoint active: http://localhost:8001
[gws] Authentication is disabled. Requests will NOT go to Google APIs.
{
  "method": "GET",
  "url": "http://localhost:8001/gmail/v1/users/me/profile"
}
```

`--dry-run` builds the request without sending it. Discovery Documents are still fetched from Google so the full command tree works — only the API request URLs are rewritten.

## Summary

Adds a `GWS_API_BASE_URL` environment variable that redirects all API requests to a custom endpoint (e.g., a mock server). When set:

- The real Discovery Document is still fetched so the CLI command tree works normally
- `root_url` and `base_url` in the Discovery Document are rewritten to the custom endpoint
- OAuth authentication is automatically skipped (mock servers don't need it)
- Works in both CLI mode and MCP server mode

## Motivation

Testing agents and skills against mock API servers is a common need for development and CI. The [`gog` CLI](https://github.com/nichochar/gog) already supports this via `GOG_ENDPOINT`. This PR brings the same capability to `gws`.

### Example usage

```bash
# Point gws at a local mock Gmail server
export GWS_API_BASE_URL=http://localhost:8099

# All commands now hit the mock server, no auth needed
gws gmail users messages list --params '{"userId": "me", "q": "is:unread"}'
```

## Changes

- **`src/discovery.rs`**: Add `custom_api_base_url()` with `LazyLock` caching; rewrite Discovery Document URLs when env var is set
- **`src/executor.rs`**: Add `resolve_auth()` that skips OAuth for custom endpoints
- **`src/main.rs`**, **`src/mcp_server.rs`**: Use `resolve_auth()` for consistent behavior

### CODE REVIEW

**Comment by jpoehnelt** [comment]
## Review Feedback

Thanks for this PR — the design is clean and smart. A few items to address before merging:

### Required

1. **Env var naming**: `GWS_API_BASE_URL` should follow the existing convention → `GOOGLE_WORKSPACE_CLI_API_BASE_URL` (consistent with `GOOGLE_WORKSPACE_CLI_TOKEN`, `GOOGLE_WORKSPACE_CLI_CREDENTIALS_FILE`, etc.)

2. **Missing changeset**: Please add a `.changeset/` file (e.g. `pnpx changeset`).

3. **Needs rebase**: The branch is behind main — `get_token()` signature has changed. Please rebase onto main.

### Recommended

4. **Add a test for `apply_base_url_override()`**: A simple unit test that constructs a `RestDescription`, calls the function, and asserts `root_url`/`base_url` were rewritten would be valuable.

5. **Security note**: Consider adding a note in the README or `AGENTS.md` that this env var should never be set in production, since it silently disables authentication.

Otherwise the approach looks solid — rewriting the Discovery Document URLs while still fetching the real schema from Google is the right call.

File: src/discovery.rs [hunk 1]
```diff
@@ -183,7 +183,27 @@ pub struct JsonSchemaProperty {
...
+
+/// Returns the custom API base URL override, if set.
+///
+/// When `GWS_API_BASE_URL` is set (e.g., `http://localhost:8099`), all API
+/// requests are directed to this endpoint instead of the real Google APIs.
+/// Authentication is skipped automatically. This is useful for testing against
+/// mock API servers.
+pub fn custom_api_base_url() -> Option<&'static str> {
+    CUSTOM_API_BASE_URL.as_deref()
+}
```
[the remark this answers is not in the document]
Comment (xdotli) [reply src/discovery.rs:199]: we added this in agents.md

**Comment by xdotli** [comment]
> .



> ## Review Feedback
> Thanks for this PR — the design is clean and smart. A few items to address before merging:
> 
> ### Required
> 1. **Env var naming**: `GWS_API_BASE_URL` should follow the existing convention → `GOOGLE_WORKSPACE_CLI_API_BASE_URL` (consistent with `GOOGLE_WORKSPACE_CLI_TOKEN`, `GOOGLE_WORKSPACE_CLI_CREDENTIALS_FILE`, etc.)
> 2. **Missing changeset**: Please add a `.changeset/` file (e.g. `pnpx changeset`).
> 3. **Needs rebase**: The branch is behind main — `get_token()` signature has changed. Please rebase onto main.
> 
> ### Recommended
> 4. **Add a test for `apply_base_url_override()`**: A simple unit test that constructs a `RestDescription`, calls the function, and asserts `root_url`/`base_url` were rewritten would be valuable.
> 5. **Security note**: Consider adding a note in the README or `AGENTS.md` that this env var should never be set in production, since it silently disables authentication.
> 
> Otherwise the approach looks solid — rewriting the Discovery Document URLs while still fetching the real schema from Google is the right call.

hi @jpoehnelt thanks for the speedy and thorough review! I have addressed all the changes and let cursor go thru it + reran my reproduce steps and it work as expected. Would appreciate another review. Thanks!

**Comment by xdotli** [comment]
@jpoehnelt thanks for the thorough review! Here's what was done to address everything:

For the base_url rewrite losing the service path — this was a real bug. The original implementation set base_url to just the custom host, which broke any API that uses a non-empty servicePath. Gmail worked fine because its servicePath is empty and the method paths include the full path (e.g. gmail/v1/users/{userId}/profile), but Drive has servicePath "drive/v3/" with short method paths like "files", so the final URL became http://localhost:8001/files instead of http://localhost:8001/drive/v3/files. Fixed by using str::replace on the original root_url so only the host gets swapped and the path portion is preserved. Added three unit tests covering both styles plus the None base_url edge case.

For the auth error logging regression — resolve_auth now returns Result so each call site can handle errors with its own context. main.rs silently falls back to unauthenticated (same as before), and mcp_server.rs restores the original "[gws mcp] Warning: Authentication failed, proceeding without credentials" message that was there before the refactor.

Also fixed cargo fmt issues that were in the original commit.

Tested all of this against a mock Gmail API server running on localhost:8001 (a FastAPI/uvicorn server implementing the full Gmail REST surface — messages, threads, labels, drafts, settings, history, profile, attachments, watch/stop, plus admin endpoints for seeding/reset/snapshots). After seeding the server, verified the following endpoints all return correct data through gws: users.getProfile, users.messages.list, users.messages.get (with path param substitution for messageId), users.labels.list, users.threads.list, users.drafts.list. Also verified Drive files.list dry-run now produces the correct URL with the service path preserved. Without the env var set, everything still hits the real Google APIs as expected.

Also changed PR title, added logging as mentioned. 

Lmk if I should start mock servers on other services and test them.

File: src/discovery.rs [hunk 2]
```diff
@@ -247,10 +263,27 @@ pub async fn fetch_discovery_document(
...
+
+fn rewrite_base_url(doc: &mut RestDescription, base: &str) {
+    let base_trimmed = base.trim_end_matches('/');
+    let new_root_url = format!("{base_trimmed}/");
+    let original_root_url = std::mem::replace(&mut doc.root_url, new_root_url);
+
+    if let Some(base_url) = &mut doc.base_url {
+        *base_url = base_url.replace(&original_root_url, &doc.root_url);
+    }
```
[the remark this answers is not in the document]
Comment (xdotli) [reply src/discovery.rs:290]: applied

File: src/discovery.rs [hunk 3]
```diff
@@ -247,10 +263,29 @@ pub async fn fetch_discovery_document(
...
+
+fn rewrite_base_url(doc: &mut RestDescription, base: &str) {
+    let base_trimmed = base.trim_end_matches('/');
+    let new_root_url = format!("{base_trimmed}/");
+    let original_root_url = std::mem::replace(&mut doc.root_url, new_root_url);
+
+    if let Some(base_url) = &mut doc.base_url {
+        if let Some(stripped_path) = base_url.strip_prefix(&original_root_url) {
+            *base_url = format!("{}{}", &doc.root_url, stripped_path);
+        }
```
[the remark this answers is not in the document]
Comment (xdotli) [reply src/discovery.rs:289]: applied

File: src/mcp_server.rs [hunk 4]
```diff
@@ -425,15 +425,7 @@ async fn handle_tools_call(params: &Value, config: &ServerConfig) -> Result<Valu
...
-    let (token, auth_method) = match crate::auth::get_token(&scopes, None).await {
-        Ok(t) => (Some(t), crate::executor::AuthMethod::OAuth),
-        Err(e) => {
-            eprintln!(
-                "[gws mcp] Warning: Authentication failed, proceeding without credentials: {e}"
-            );
-            (None, crate::executor::AuthMethod::None)
-        }
-    };
+    let (token, auth_method) = crate::executor::resolve_auth(&scopes, None).await;
```
[the remark this answers is not in the document]
Comment (xdotli) [reply src/mcp_server.rs]: applied in a recent commit

File: src/main.rs [hunk 5]
```diff
@@ -235,11 +235,9 @@ async fn run() -> Result<(), GwsError> {
...
     let scopes: Vec<&str> = method.scopes.iter().map(|s| s.as_str()).collect();
 
-    // Authenticate: try OAuth, otherwise proceed unauthenticated
-    let (token, auth_method) = match auth::get_token(&scopes, account.as_deref()).await {
-        Ok(t) => (Some(t), executor::AuthMethod::OAuth),
-        Err(_) => (None, executor::AuthMethod::None),
-    };
+    // Authenticate: skips OAuth automatically when GOOGLE_WORKSPACE_CLI_API_BASE_URL is set.
+    let (token, auth_method) =
+        executor::resolve_auth(&scopes, account.as_deref()).await;
```
[the remark this answers is not in the document]
Comment (xdotli) [reply src/main.rs]: applied

File: src/executor.rs [hunk 6]
```diff
@@ -38,6 +38,21 @@ pub enum AuthMethod {
...
+/// that check so every call-site doesn't need to know about the env var.
+pub async fn resolve_auth(scopes: &[&str], account: Option<&str>) -> (Option<String>, AuthMethod) {
+    if crate::discovery::custom_api_base_url().is_some() {
+        return (None, AuthMethod::None);
+    }
+    match crate::auth::get_token(scopes, account).await {
+        Ok(t) => (Some(t), AuthMethod::OAuth),
+        Err(_) => (None, AuthMethod::None),
+    }
+}
```
[the remark this answers is not in the document]
Comment (xdotli) [reply src/executor.rs]: addressed

File: src/discovery.rs [hunk 7]
```diff
@@ -323,4 +371,35 @@ mod tests {
...
+            version: "v3".to_string(),
+            root_url: "https://www.googleapis.com/".to_string(),
+            base_url: Some("https://www.googleapis.com/drive/v3/".to_string()),
+            ..Default::default()
+        };
+
+        rewrite_base_url(&mut doc, "http://localhost:8099/");
+        assert_eq!(doc.root_url, "http://localhost:8099/");
+        assert_eq!(doc.base_url.as_deref(), Some("http://localhost:8099/"));
+    }
```
[the remark this answers is not in the document]
Comment (xdotli) [reply src/discovery.rs:431]: addressed

**Comment by xdotli** [comment]
hi @jpoehnelt, I applied all gemini's suggestions and resolved merge conflicts. I reviewed every file again myself + dry run locally + used our gmail mock server to test. everything works as expected. would be great to get a rereview when you got the time. much appreciated!

**Comment by xdotli** [comment]
how do i reopen it?
